src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/RunTestsHandler.cs

```diff
@@ -0,0 +1,131 @@
+    {
+        Contract.ThrowIfNull(context.Document);
+        using var progress = BufferedProgress.Create(request.PartialResultToken);
+
+        // First, build to make sure we have a relatively up to date project.
+        await BuildAsync(context.Document, progress, cancellationToken);
+
+        var projectOutputPath = context.Document.Project.OutputFilePath;
+        Contract.ThrowIfFalse(File.Exists(projectOutputPath), $"Output path {projectOutputPath} is missing");
+
+        // Find the appropriate vstest.console.dll from the SDK.
+        var vsTestConsolePath = await dotnetCliHelper.GetVsTestConsolePathAsync(cancellationToken);
+
+        // Instantiate the test platform wrapper.
+        var vsTestConsoleWrapper = new VsTestConsoleWrapper(vsTestConsolePath, new ConsoleParameters
+        {
+            LogFilePath = Path.Combine(serverConfiguration.ExtensionLogDirectory, "testLogs", "vsTestLogs.txt"),
+            TraceLevel = GetTraceLevel(serverConfiguration),
+            EnvironmentVariables = new()
+            {
+                // Reset dotnet root to the user's original configuration so that vs test console can find the right runtimes.
+                { DotnetCliHelper.DotnetRootEnvVar, DotnetCliHelper.GetUserDotnetRoot() },
+            }
+        });
+
+        var testCases = await testDiscoverer.DiscoverTestsAsync(request.Range, context.Document, projectOutputPath, progress, vsTestConsoleWrapper, cancellationToken);
+        if (testCases.IsEmpty)
+        {
+            return progress.GetValues() ?? Array.Empty<RunTestsPartialResult>();
+        }
+
+        await testRunner.RunTestsAsync(testCases, progress, vsTestConsoleWrapper, cancellationToken);
+
+        return progress.GetValues() ?? Array.Empty<RunTestsPartialResult>();
+    }
+
+    /// <summary>
+    /// Shells out to the .NET CLI to build the project to ensure that its relatively up to date as of this request.
+    /// We can never be entirely sure that this build happens on exactly the same snapshot that this request was made on
+    /// as something could have changed on the client side immediately after this request was triggered.
+    ///
+    /// However if we don't do a build it is likely the user code is very different from what they last built
+    /// which results in a confusing experience.
+    /// </summary>
+    private async Task BuildAsync(Document document, BufferedProgress<RunTestsPartialResult> progress, CancellationToken cancellationToken)
+    {
+        var workingDirectory = Path.GetDirectoryName(document.Project.FilePath);
+        Contract.ThrowIfNull(workingDirectory, $"Unable to get working directory for project {document.Project.Name}");
+
+        // TODO - we likely need to pass the no-restore flag once we have automatic restore enabled.
+        // https://github.com/dotnet/vscode-csharp/issues/5725
+        var arguments = "build";
```

Comment:
this approach seems odd.  is tehre no LSP way to invoke the build?  what happens when VScode itself is trying to build?  Shouldn't we defer to that, and then do the next steps after that happens?

Reply:
There is not an LSP way to invoke a build.  It is possible to invoke the vscode build task (as defined in a project's launch.json) from the vscode client, but that build task isn't necessarily what we want either.  We want to build this test project specifically, we don't need to build anything else.  The users build task may not build this project, or it may build an entire solution unnecessarily.

This is the best way to get a minimal build here (this replicates O# behavior as well).